### PR TITLE
WeaponDesign serializer

### LINKED ISSUE #333: Create WeaponDesignBinaryPackage
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

Create WeaponDesignBinaryPackage and associated unit test.

## Intended Design

N/A

## Location

Work in a new branch based from [taverns-serialization](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/taverns-serialization)

## Related Issues

Blocked by
- #336

## Requirements

N/A

## Additional information

See [StaticBodyPropertiesBinaryPackage](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/taverns-serialization/source/GameInterface/Serialization/Impl/StaticBodyPropertiesBinaryPackage.cs) and it's [test](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/taverns-serialization/source/GameInterface.Tests/Serialization/SerializerTests/StaticBodyPropertiesSerializationTest.cs)

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### LINKED ISSUE #408: Create PieceDataBinaryPackage
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

Create PieceDataBinaryPackage and associated unit test.

## Intended Design

N/A

## Location

Work in a new branch based from [taverns-serialization](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/taverns-serialization)

## Related Issues

N/A

## Requirements

N/A

## Additional information

N/A

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the new behaviour?

Resolves #333 

## Other information
Had to add PieceData (which does not have a test yet) in order for the WeaponDesign to work

### CODE REVIEW

File: source/GameInterface/Serialization/Impl/CraftingTemplateBinaryPackage.cs [hunk 1]
```diff
@@ -15,11 +17,21 @@ public CraftingTemplateBinaryPackage(CraftingTemplate obj, BinaryPackageFactory
         }
         public override void Pack()
         {
-            templateId = Object.StringId;
```
Comment (garrettluskey) [inline source/GameInterface/Serialization/Impl/CraftingTemplateBinaryPackage.cs]: Revert these changes and in your test do something along the lines of https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/taverns-serialization/source/GameInterface.Tests/Serialization/SerializerTests/ItemRosterSerializationTest.cs

File: source/GameInterface.Tests/GameInterface.Tests.csproj [hunk 2]
```diff
@@ -94,6 +94,7 @@
     <Reference Include="System.Threading.Tasks.Extensions, Version=4.2.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
       <HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.4\lib\net461\System.Threading.Tasks.Extensions.dll</HintPath>
     </Reference>
+    <Reference Include="System.Xml" />
```
Comment (garrettluskey) [inline source/GameInterface.Tests/GameInterface.Tests.csproj]: Not sure why this is included, remove this if possible

File: source/GameInterface/Serialization/Impl/PieceDataBinaryPackage.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.Core;
+using TaleWorlds.ObjectSystem;
+
+namespace GameInterface.Serialization.Impl
+{
+    [Serializable]
+    public class PieceDataBinaryPackage : BinaryPackageBase<PieceData>
```
Comment (garrettluskey) [inline source/GameInterface/Serialization/Impl/PieceDataBinaryPackage.cs:9]: Create test for this as per #408

File: source/GameInterface/Serialization/Impl/WeaponDesignBinaryPackage.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    [Serializable]
+    public class WeaponDesignBinaryPackage : BinaryPackageBase<WeaponDesign>
+    {
+        public WeaponDesignBinaryPackage(WeaponDesign obj, BinaryPackageFactory binaryPackageFactory) : base(obj, binaryPackageFactory)
+        {
+        }
+
+        public override void Pack()
+        {
+            foreach (FieldInfo field in ObjectType.GetAllInstanceFields())
```
Comment (garrettluskey) [inline source/GameInterface/Serialization/Impl/WeaponDesignBinaryPackage.cs:17]: There is some cached data that should be excluded i.e _cachedHashedCodeInt

File: source/GameInterface.Tests/Serialization/SerializerTests/WeaponDesignSerializationTest.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            object obj = BinaryFormatterSerializer.Deserialize(bytes);
+
+            Assert.IsType<WeaponDesignBinaryPackage>(obj);
+
+            WeaponDesignBinaryPackage returnedPackage = (WeaponDesignBinaryPackage)obj;
+
+            WeaponDesign newWeaponDesign = returnedPackage.Unpack<WeaponDesign>();
+
+            Assert.Equal(WeaponDesign.CraftedWeaponLength, newWeaponDesign.CraftedWeaponLength);
+            Assert.Equal(WeaponDesign.HolsterShiftAmount, newWeaponDesign.HolsterShiftAmount);
```
Comment (garrettluskey) [inline source/GameInterface.Tests/Serialization/SerializerTests/WeaponDesignSerializationTest.cs:89]: Add assertions for Template and _usedPieces fields

**Review by garrettluskey — CHANGES_REQUESTED** [review]
A few things need to be updated, great progress thanks again

**Comment by EgardA** [comment]
Updated most things but I could not get the Assert to work with the UsedPieces Array for some reason (literally the same?)

File: source/GameInterface/Serialization/Impl/WeaponDesignBinaryPackage.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            }
+        }
+
+        protected override void UnpackInternal()
+        {
+            TypedReference reference = __makeref(Object);
+            foreach (FieldInfo field in StoredFields.Keys)
+            {
+                field.SetValueDirect(reference, StoredFields[field].Unpack());
+            }
```
Comment (garrettluskey) [inline source/GameInterface/Serialization/Impl/WeaponDesignBinaryPackage.cs:31]: Call `BuildHashedCode` after the foreach loop

**Review by garrettluskey — APPROVED** [review]

File: source/GameInterface/Serialization/Impl/WeaponDesignBinaryPackage.cs (hunk 6 above)
Comment (garrettluskey) [reply source/GameInterface/Serialization/Impl/WeaponDesignBinaryPackage.cs:31]: It's less expensive to store the data rather than recalculate
